File: github_code/red-black-tree/java/102.java

public class redblacktree {

	  private treeNode Root;
	  
	  public treeNode getRoot() { return Root; }
	  
	  // Constructs a new empty red-black tree.
	  public redblacktree() { Root = Root.nullnode;}
	  public void clear() { Root = Root.nullnode; }
	  
	  public boolean isEmpty() { return Root.isEmpty(); }	  

	  public int size() { return Root.size(); }
	    
	  public treeNode find(int key) {
		  if (Root.isEmpty()) return Root;
		  return Root.find(key);
	  }

	  public void checkRedBlack() {
		  if (Root.isRed) 
			  System.out.println( "The root is red");
		  Root.bheight();
	  }
	  
	  public void insert(int id) { // call recursive insert
	       treeNode nnd = new treeNode(id);    // make new node

	       if (Root.isEmpty()) Root = nnd;
	       else if (id < Root.iData) {
	    	   if (Root.left.isEmpty())
	    		   Root.left = nnd;
	    	   else if (insert(nnd, Root.left, Root)) {
	  	    	   // red-red violation exists at Root.left and its children
			       if (Root.right.isRed) { // red uncle case:
			    	   Root.right.isRed = Root.left.isRed = false;
			       } else { // black uncle case:			    	   
			    	   Root.isRed = true;
			           if (nnd.iData < Root.left.iData) {
			               Root = Root.rotateToRight();
			           } else {
			               Root = Root.doubleRotateToRight();
			           } 			 
			       } // else
	    	   }
	       } else if (Root.right.isEmpty()) {
	    		   Root.right = nnd;
	       } else if (insert(nnd, Root.right, Root)) {
  	    	   // red-red violation exists at Root.right and its children
		       if (Root.left.isRed) { // red uncle case:
		    	   Root.right.isRed = Root.left.isRed = false;
		       } else { // black uncle case:		    	   
		    	   Root.isRed = true;
		           if (nnd.iData >= Root.right.iData) {
		               Root = Root.rotateToLeft();
		           } else {
		               Root = Root.doubleRotateToLeft();
		           } 
		       } // else
	       }
	       Root.isRed = false;
	    }  // end insert()

	public void remove(int id){
		remove(id, true);

	}

    public void remove(int id, boolean delete) {
    	treeNode node = Root.find(id);
    	if (!node.isEmpty()){
    		treeNode parent = parentOf(node);
    		boolean leftChild = (parent.left==node);
    		treeNode sibling = leftChild ? parent.right : parent.left;
    		if(node.right.isEmpty() && node.left.isEmpty()) {  // no child
    			if (node.isRed){  // node is red
    				if (delete) if (leftChild) parent.left = Root.nullnode; else parent.right = Root.nullnode;
    			}else{  //node is black
    				if (sibling.isEmpty() && parent.isRed){ // no sibling and red parent
    					if (delete) if (leftChild) parent.left = Root.nullnode; else parent.right = Root.nullnode;
    					parent.isRed = false;
    				}else{ //has sibling
	    				if (sibling.isRed){  // case 1
	    					sibling.isRed = false;
	    					parent.isRed = true;
	    					if (leftChild) leftRotate(parent); else rightRotate(parent);
	    					remove(id);
	    				}else{ 
	    					treeNode distant_nephew = leftChild ? sibling.right : sibling.left;
	    					treeNode close_nephew = leftChild ? sibling.left : sibling.right;
	    					if(distant_nephew.isRed){  // case 2
	    						sibling.isRed = parent.isRed;
	    						parent.isRed = false;
	    						distant_nephew.isRed = false;
	    						if (delete) if (leftChild) parent.left = Root.nullnode; else parent.right = Root.nullnode;
	    						if (leftChild) leftRotate(parent); else rightRotate(parent);
	    					}else if(close_nephew.isRed){   // case 3
	    						close_nephew.isRed = false;
	    						sibling.isRed = true;
	    						if (leftChild) rightRotate(parent); else leftRotate(parent);
	    						remove(id);
	    					}else if(sibling.left.isEmpty() && sibling.right.isEmpty()){ 
	    						if(parent.isRed) {        // case 4
		    						parent.isRed = false;
		    						sibling.isRed = true;
		    						if (delete) if (leftChild) parent.left = Root.nullnode; else parent.right = Root.nullnode;
		    						sibling.isRed = true;
		    						if (delete){
		    							if (leftChild) parent.left = Root.nullnode; else parent.right = Root.nullnode;
		    							remove(parent.iData, false);
		    						}
		    					}
	    					}
	    				}
    				}				
    			}
    		}else if(node.right.isEmpty()^node.left.isEmpty()){ // one child (it will always be black-red)
    				treeNode child = (node.right.isEmpty() ? node.right : node.left);
	    			child.isRed = false;
	    			node = child;
    		}else{                             // two child
	    		treeNode minNode = node.right.findMin();
	    		remove(minNode.iData);
		    	node.iData = minNode.iData;
		    }
	    }
    }

	  /* *************************************************** *
	   *  PRIVATE METHODS                                    *
	   * *************************************************** */

	  	private void leftRotate(treeNode parent){
	  		int temp = parent.iData;
	  		treeNode a = parent.left;
	  		treeNode b = parent.right.left;
	  		treeNode c = parent.right.right;
	  		parent.iData = parent.right.iData;
	  		parent.right.iData = temp;
	  		parent.left = parent.right;
	  		parent.left.left = a;
	  		parent.left.right = b;
	  		parent.right = c;
	  	}

	  	private void rightRotate(treeNode parent){
	  		int temp = parent.iData;
	  		treeNode a = parent.left.left;
	  		treeNode b = parent.left.right;
	  		treeNode c = parent.right;
	  		parent.iData = parent.left.iData;
	  		parent.left.iData = temp;
	  		parent.right = parent.left;
	  		parent.left = a;
	  		parent.right.left = b;
	  		parent.right.right = c;
	  	}

	  	private treeNode parentOf(treeNode node) {
    		if (Root.iData == node.iData) return Root.nullnode;
    		treeNode parent = Root.nullnode;
    		treeNode current = Root;
    		while(current!=Root.nullnode){
    			if (node.iData < current.iData){
    				parent = current;
    				current = current.left;
    			}else if (node.iData > current.iData){
    				parent = current;
    				current = current.right;
    			}else{
    				break;
    			}
    		}
    		return parent;
    	}

	  /* Inserts a node into tree and returns a boolean */
	  private boolean insert(treeNode nnd, treeNode t, treeNode par) {
		  // return true iff t is red and t has a red child

		  if (nnd.iData < t.iData) {
		      if (t.left.isEmpty()) {
		         t.left = nnd;  //attach new node as leaf
		      } else if (insert(nnd, t.left, t)) { 
		    	 // red-red violation exists at t.left and its children
		    	 if (t.right.isRed) { // red uncle case:
		    		 t.right.isRed = t.left.isRed = false;
		    		 t.isRed = true;
		    	 } else { // black uncle case:
		    		 treeNode nt;
		    		 
		             if (nnd.iData < t.left.iData) {
		                nt = t.rotateToRight();
		             } else {
		                nt = t.doubleRotateToRight();
		             } 
		    		 t.isRed = true;
		    		 nt.isRed = false;
		    		 if (nt.iData < par.iData) par.left = nt;
		    		 else par.right = nt;
		         } //if
		      }
		      
		      return (t.isRed && t.left.isRed);
		 } else { // branch right
		      if (t.right.isEmpty()) {
		           t.right = nnd;   // attach new node as leaf
			  } else if (insert(nnd, t.right, t)) { 
			       // red-red violation exists at t.right and its children
			       if (t.left.isRed) { // red uncle case:
			    		 t.right.isRed = t.left.isRed = false;
			    		 t.isRed = true;
			       } else { // black uncle case:
			    		 treeNode nt;
			    		 
			             if (nnd.iData >= t.right.iData) {
			                nt = t.rotateToLeft();
			             } else {
			                nt = t.doubleRotateToLeft();
			             } 
			    		 t.isRed = true;
			    		 nt.isRed = false;
			    		 if (nt.iData < par.iData) par.left = nt;
			    		 else par.right = nt;
			       } // else
			  } 
		      
			  return (t.isRed && t.right.isRed);
		 } // else
	  }  // end insert


}

File: github_code/red-black-tree/cpp/102.c
#ifdef _REDBLACKTREE_H_
#include<stdexcept>

#include <iostream>
#include <math.h>
//W
//private

// recursive helper function for deep copy
// creates a new node "thisnode" based on sourcenode's contents, links back to parentnode,
//   and recurses to create left and right children
template <class T>
Node<T>* RedBlackTree<T>::CopyTree(Node<T>* thisnode, Node<T>* sourcenode, Node<T>* parentnode)
{
	if(sourcenode != NULL)
	{
		if(thisnode == NULL)
		{
			thisnode = new Node<T>();
		}
		thisnode->data = sourcenode->data; //creates a new node
		thisnode->is_black = sourcenode->is_black; //sets node's color

		thisnode = parentnode;

		if(sourcenode->left != NULL)
		{
			CopyTree( new Node<T>(),sourcenode->left,thisnode);
		}
		if(sourcenode->right != NULL)
		{
			CopyTree( new Node<T>(),sourcenode->right,thisnode);
		}
	}
}


// recursive helper function for tree deletion
// deallocates nodes in post-order
template <class T>
void RedBlackTree<T>::RemoveAll(Node<T>* node)
{
	//Postorder traversal
	if(node != NULL)
	{
		RemoveAll(node->left);
		RemoveAll(node->right);	
		delete node;
		size--;
	}
}
// hard!
// Tree fix, performed after removal of a black node
// Note that the parameter x may be NULL
template <class T>
void RedBlackTree<T>::RBDeleteFixUp(Node<T>* x, Node<T>* xparent, bool xisleftchild)
{
	bool isBlack = false; //go into while loop with x = NULL
	Node<T>* w = NULL;
	if (x == NULL || x->is_black) //null = black node
	{
		isBlack = true;
	}

	while (x != root && isBlack)
	{
		if (xisleftchild)//x is leftchild
		{
			w = xparent->right; //w = x's sibling
			if (!w->is_black)
			{
				w->is_black = true;
				xparent->is_black = false;
				LeftRotate(xparent); //rotate around x's parent
				w = xparent->right;
			}
			if ((w->left == NULL || w->left->is_black) && (w->right == NULL || w->right->is_black)) //performing on NULL crashes
			{
				w->is_black = false;
				x = xparent;
				xparent = x->p;
				xisleftchild = (xparent != NULL && x == xparent->left);
			}
			else {
				if (w->right == NULL || w->right->is_black)
				{
					w->left->is_black = true;
					w->is_black = false;
					RightRotate(w);
					w = xparent->right;
				}
				w->is_black = xparent->is_black;
				xparent->is_black = true;
				if (w->right != NULL)
				{
					w->right->is_black = true;
				}
				LeftRotate(xparent);
				x = root;
				xparent = NULL;
			}
		}
		else //swap left and right
		{
			w = xparent->left;
			if (!w->is_black)
			{
				w->is_black = true;
				xparent->is_black = false;
				RightRotate(xparent);
				w = xparent->left;
			}
			if ((w->left == NULL || w->left->is_black) && (w->right == NULL || w->right->is_black))
			{
				w->is_black = false;
				x = xparent;
				xparent = x->p;
				xisleftchild = (xparent != NULL && x == xparent->left);
			}
			else
			{
				if (w->left == NULL || w->left->is_black)
				{
					w->right->is_black = true;
					w->is_black = false;
					LeftRotate(w);
					w = xparent->left;
				}
				w->is_black = xparent->is_black;
				xparent->is_black = true;
				if (w->left != NULL)
				{
					w->left->is_black = true;
				}
				RightRotate(xparent);
				x = root;
				xparent = NULL;
			}
		}
	}
	if (x != NULL) //performing is_black on NULL errors
	{
		x->is_black = true;
	}
}

// Calculates the height of the tree
// Requires a traversal of the tree, O(n)
template <class T>
unsigned int RedBlackTree<T>::CalculateHeight(Node<T>* node) const
{
	if(node == NULL) //empty so height of 0
	{
		return 0;
	}
	else
	{
		unsigned int tempLeft = 0; //count the left children
		unsigned int tempRight = 0; //count the right children

		if(node->left != NULL)
		{
			unsigned int tempLeft = CalculateHeight(node->left);
		}
		if(node->right != NULL)
		{
			unsigned int tempRight = CalculateHeight(node->right);
		}

		if(tempLeft > tempRight) //compares which one is more
		{
			return tempLeft + 1;
		}
		else
		{
			return tempRight + 1;
		}
	}

}

//public

// default constructor
template <class T>
RedBlackTree<T>::RedBlackTree()
{
	size = 0;
	root = NULL;
}

// copy constructor, performs deep copy of parameter
template <class T>
RedBlackTree<T>::RedBlackTree(const RedBlackTree<T>& rbtree)
{
	root = new Node<T>(root->data); 
	CopyTree(root, rbtree.GetRoot(), NULL);	
}

// destructor
// Must deallocate memory associated with all nodes in tree
template <class T>
RedBlackTree<T>::~RedBlackTree()
{
	RemoveAll();
}

// overloaded assignment operator
template <class T>
RedBlackTree<T>& RedBlackTree<T>::operator=(const RedBlackTree<T>& rbtree)
{
	if (this != &rbtree) 
	{
		this->RemoveAll(root);
		CopyTree(NULL, rbtree.root, NULL);
		size = rbtree.size;
	}
	return *this;
}

// Accessor functions

// Calls BSTInsert and then performs any necessary tree fixing.
// If item already exists, do not insert and return false.
// Otherwise, insert, increment size, and return true.
template <class T>
bool RedBlackTree<T>::Insert(T item)
{
	//performs if item is not found in tree
	if(Search(item) == false)  
	{
		Node<T>* x = BSTInsert(item);  //calls BSTinsert and sets to x
		Node<T>* y = NULL;
		x->is_black = false;  //defaults to black node
		while(x->p != NULL && x->p->is_black == false) //iterates until root or black parent reached
		{
			if(x->p == x->p->p->left)
			{
				y = x->p->p->right; //uncle of x
				if(y != NULL && y->is_black == false) //NULL means black 
				{
					x->p->is_black = true;
					y->is_black = true;
					x->p->p->is_black = false;
					x = x->p->p;
				}
				else
				{
					if(x == x->p->right)
					{
						x = x->p;
						LeftRotate(x);
					}
					x->p->is_black = true;
					x->p->p->is_black = false;
					RightRotate(x->p->p);
				}
			}
			else //x->p == x->p->p->right
			{
				y = x->p->p->left;
				if(y != NULL && y->is_black == false)
				{
					x->p->is_black = true;
					y->is_black = true;
					x->p->p->is_black = false;
					x = x->p->p;
				}
				else
				{
					if(x == x->p->left)
					{
						x = x->p;
						RightRotate(x);
					}

					x->p->is_black = true;
					x->p->p->is_black = false;					
					LeftRotate(x->p->p);

				}
			}

		}
		root->is_black = true;
		size++; //increment size by 1
		return true;
	}
	return false;
}

// Removal of an item from the tree.
// Must deallocate deleted node after RBDeleteFixUp returns
template <class T>
bool RedBlackTree<T>::Remove(T item)
{
	if(Search(item) == false) //returns false if Search does not find item
	{
		return false;
	}

	Node<T>* z = root;

	while (z != NULL) //binary search through tree for item
	{
		if (item == z->data)
			break;
		else if (item < z->data)
			z = z->left;
		else
			z = z->right;
	}

	Node<T>* y = NULL;
	if (z->left == NULL || z->right == NULL)
	{
		y = z; //y is the node to be removed
	}
	else
	{
		y = Predecessor(z); //sets y to predecessor
	}

	Node<T>* x = NULL;
	if (y->left != NULL)
	{
		x = y->left;
	}
	else
	{
		x = y->right;
	}
	if (x != NULL)
	{
		x->p = y->p;
	}
	Node<T>* xParent = y->p;
	bool yLCH = false;

	if (y->p == NULL) //x is root
	{
		root = x;
	}
	else if (y == y->p->left) //y is predecessor
	{
		y->p->left = x;
		yLCH = true;
	}
	else
	{
		y->p->right = x;
		yLCH = false;
	}

	if (y != z)
	{
		z->data = y->data;
	}

	if (y->is_black)
	{
		RBDeleteFixUp(x, xParent, yLCH);
	}

	delete y; //DELETE Y AFTER
	size--; //decrement
	z = NULL;
	return true;
}

// deletes all nodes in the tree. Calls recursive helper function.
template <class T>
void RedBlackTree<T>::RemoveAll()
{
	RemoveAll(root);
	root = NULL;
}

// returns the number of items in the tree
template <class T>
unsigned int RedBlackTree<T>::Size() const
{
	return size;
}

// returns the height of the tree, from root to deepest null child. Calls recursive helper function.
// Note that an empty tree should have a height of 0, and a tree with only one node will have a height of 1.
template <class T>
unsigned int RedBlackTree<T>::Height() const
{
	return CalculateHeight(root);
}


#endif